File: w2/readability/readability.c
#include <cs50.h>
#include <ctype.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

int countLetters(string text);
int countWords(string text);
int countSentence(string text);
int calcIndex(int letters, int words, int sentences);

int main(void)
{
    string text = get_string("Text: ");

    int letters = countLetters(text);
    int words = countWords(text);
    int sentence = countSentence(text);

    float index = calcIndex(letters, words, sentence);

    if (index < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (index > 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %i\n", (int) index);
    }
}

int countLetters(string text)
{
    int letters = 0;
    for (int i = 0, length = strlen(text); i < length; i++)
    {
        if (isalpha(text[i]))
        {
            letters++;
        }
    }
    return letters;
}

int countWords(string text)
{
    int wordCount = 0;
    for (int i = 0, length = strlen(text); i < length; i++)
    {
        if (!isspace(text[i]) && (isspace(text[i + 1]) || text[i + 1] == '\0'))
        {
            wordCount++;
        }
    }
    return wordCount;
}

int countSentence(string text)
{
    int sentences = 0;
    for (int i = 0, length = strlen(text); i < length; i++)
    {
        if (text[i] == '.' || text[i] == '!' || text[i] == '?')
        {
            sentences++;
        }
    }
    return sentences;
}

int calcIndex(int letters, int words, int sentences)
{
    float L = (float) letters / words * 100;
    float S = (float) sentences / words * 100;
    return round(0.0588 * L - 0.296 * S - 15.8);
}

File: w6/readability.py
from cs50 import get_string


def main():
    text = get_string("Text: ")
    letters = countLetters(text)
    words = countWords(text)
    sentences = countSentences(text)
    index = calcIndex(letters, words, sentences)

    if index < 1:
        print("Before Grade 1")
    elif index > 16:
        print("Grade 16+")
    else:
        print(f"Grade {index}")


def countLetters(text):
    letters = 0
    for char in text:
        if char.isalpha():
            letters += 1
    return letters


def countWords(text):
    words = 0
    length = len(text)
    for i in range(length):
        if not text[i].isspace() and (i + 1 == length or text[i + 1].isspace()):
            words += 1
    return words


def countSentences(text):
    sentences = 0
    for char in text:
        if char in {".", "!", "?"}:
            sentences += 1
    return sentences


def calcIndex(letters, words, sentences):
    L = letters / words * 100
    S = sentences / words * 100
    index = round(0.0588 * L - 0.296 * S - 15.8)
    return index


main()
